aten/src/ATen/native/cuda/TensorTopK.cu

```diff
@@ -0,0 +1,372 @@
+      inputSlices *= inputInfo.sizes[i];                                  \
+    }                                                                     \
+    int64_t topKSlices = 1;                                               \
+    for (int i = 0; i < topKInfo.dims; ++i) {                             \
+      topKSlices *= topKInfo.sizes[i];                                    \
+    }                                                                     \
+                                                                          \
+    dim3 grid;                                                            \
+    if (!THC_getGridFromTiles(inputSlices, grid)) {                       \
+      THError("Slice to sort is too large");                              \
+    }                                                                     \
+                                                                          \
+    dim3 block(std::min(THCRoundUp(sliceSize, (int64_t) C10_WARP_SIZE), (int64_t) 1024)); \
+                                                                          \
+    /* This is used as a template parameter to calculate indices. */      \
+    /* We only specialize it if all collapsed dim sizes are the */        \
+    /* same; otherwise, we use -1 which is the specialization */          \
+    /* parameter for arbitrary dimensions */                              \
+    int allDims = inputInfo.dims;                                         \
+    if (topKInfo.dims != allDims || indicesInfo.dims != allDims) {        \
+      allDims = -1;                                                       \
+    }                                                                     \
+                                                                          \
+    RUN_DIM(INDEX_T);                                                     \
+  });
+
+  // the below is safe with 0-dimensional tensors because it is based on
+  // TensorInfo which implicitly expands to 1-dimensional.
+  if (input.numel() > 0) {
+    // Based on required index size, run the algorithm with the
+    // appropriate index type
+    if (at::cuda::detail::canUse32BitIndexMath(input) &&
+        at::cuda::detail::canUse32BitIndexMath(values) &&
+        at::cuda::detail::canUse32BitIndexMath(indices)) {
+      RUN_T(uint32_t);
+    } else {
+      RUN_T(uint64_t);
+    }
+  }
+#undef RUN_T
+#undef RUN_DIM
+#undef RUN_DIR
+#undef RUN_K
+
+  // Sort the results if the user wants them sorted, since our
+  // selection routine does not ensure sorting
+  if (sorted && values.numel() > 1) {
+    // FIXME: the k/v inplace sort along slice only works for size <=
+    // 2048 at the moment
+    // Workaround:
+    // CUDA 8 uses more shared memory than 7.5 for bitonicSortKVInPlace,
+    // and so for the double word types,
+    // we get "too many resources requested for launch" in the 2048 case
+#if CUDA_VERSION >= 8000
```

Comment:
CUDA 8.x and 9.x are no longer supported, so we can unconditionally assume `CUDA_VERSION >= 8000` always true